dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/DefaultProgramRuleEntityMapperService.java

```diff
@@ -545,15 +612,29 @@ private String getEventDataValue( EventDataValue dataValue )
         return dataValue.getValue() != null ? dataValue.getValue() : "";
     }
 
-    private ValueType getValueTypeForDataElement( String dataElementUid )
+    private ValueType getValueTypeForDataElement( String identifier, IdScheme idScheme )
     {
-        return dataElementToValueTypeCache.get( dataElementUid, () -> {
-            DataElement dataElement = dataElementService.getDataElement( dataElementUid );
+        return dataElementToValueTypeCache.get( identifier, () -> {
+            // TODO(DHIS2-12282) support idScheme ATTRIBUTE
+            DataElement dataElement = null;
+            if ( idScheme.getIdentifiableProperty() == IdentifiableProperty.CODE )
+            {
+                dataElement = identifiableObjectManager.getByCode( DataElement.class, identifier );
+            }
+            else if ( idScheme.getIdentifiableProperty() == IdentifiableProperty.NAME )
+            {
+                dataElement = identifiableObjectManager.getByName( DataElement.class, identifier );
+            }
+            else
+            { // TODO(DHIS-12282) ok to default to idScheme UID if an
```

Comment:
Not sure it should be possible to end up with NULL or something else, it should always be UID, CODE, ATTRIBUTE, AUTO, etc. UID is default, and AUTO will also be UID. If for some reason we do get null or something not a idScheme, we should throw, not fallback, because then our assumption is fundamentally wrong imo

Reply:
So now the program rule engine will accept ID, CODE and NAME, if IdScheme is other than these three then it will raise an exception. I think this is worth mentioning in the documentation.